File: solutions/c/protein-translation/1/protein_translation.c
#include "protein_translation.h"
#include <string.h>

typedef struct {
    const char code[4];
    amino_acid_t aminoacid;
} k_map_t;

protein_t protein(const char *const rna) {
    protein_t p = { .valid = true, .count = 0 };
    
    if (!rna) {
        p.valid = false;
        return p;
    }
    
    const k_map_t map[] = {
        {"AUG", Methionine},
        {"UUU", Phenylalanine}, {"UUC", Phenylalanine},
        {"UUA", Leucine}, {"UUG", Leucine},
        {"UCU", Serine}, {"UCC", Serine}, {"UCA", Serine}, {"UCG", Serine},
        {"UAU", Tyrosine}, {"UAC", Tyrosine},
        {"UGU", Cysteine}, {"UGC", Cysteine},
        {"UGG", Tryptophan},
        {"UAA", Stop}, {"UAG", Stop}, {"UGA", Stop}
    };

    size_t map_size = sizeof(map) / sizeof(map[0]);
    
    for (size_t i = 0; rna[i] != '\0'; i += 3) {
        bool found = false;

        if (rna[i+1] == '\0' || rna[i+2] == '\0') {
            p.valid = false;
            break;
        }
        
        for (size_t j = 0; j < map_size; j++) {
            if (!strncmp(&rna[i], map[j].code, 3)) {
                if (map[j].aminoacid == Stop) {
                    return p;
                }

                p.amino_acids[p.count++] = map[j].aminoacid;
                found = true;
                break; 
            }
        }
        if (!found) {
            p.valid = false;
            break;
        }
    }
    return p;
}

File: solutions/ruby/protein-translation/1/protein_translation.rb
class InvalidCodonError < ArgumentError; end

module Translation
  POTEIN = {
      "Methionine"    => %w[AUG],
      "Phenylalanine" => %w[UUU UUC],
      "Leucine"       => %w[UUA UUG],
      "Serine"        => %w[UCU UCC UCA UCG],
      "Tyrosine"      => %w[UAU UAC],
      "Cysteine"      => %w[UGU UGC],
      "Tryptophan"    => %w[UGG],
      "Stop"          => %w[UAA UAG UGA]
    }

    P_MAP = POTEIN.each_with_object({}) do |(name, code), hash|
      code.each { |c| hash[c] = name }
    end
  
  def self.of_rna(strand)
    rna = strand.scan(/.{1,3}/)

    return [] if strand.empty?

    rna.take_while { |r| P_MAP[r] != "Stop" }.map { |r| P_MAP[r] || raise(InvalidCodonError) }
  end
end